### PR TITLE
Add filter to generate the voxel cell centers for an Image Geometry

### PR DESCRIPTION
**Pull request by imikejackson**
Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

### CODE REVIEW

File: Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+//
+// -----------------------------------------------------------------------------
+void GenerateVertexCoordinates::setupFilterParameters()
+{
+  FilterParameterVectorType parameters;
+  {
+    DataContainerSelectionFilterParameter::RequirementType req;
+    req.dcGeometryTypes = IGeometry::Types(2);
+    req.dcGeometryTypes[0] = IGeometry::Type::Image;
+    req.dcGeometryTypes[0] = IGeometry::Type::RectGrid;
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp]:
```suggestion
    req.dcGeometryTypes = {IGeometry::Type::Image, IGeometry::Type::RectGrid};
```

File: Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+  }
+
+  DataContainer::Pointer dc = getDataContainerArray()->getPrereqDataContainer(this, getSelectedDataContainerName());
+  if(getErrorCode() < 0)
+  {
+    return;
+  }
+
+  IGeometry::Pointer fromGeometry = dc->getGeometry();
+  if(nullptr == fromGeometry.get())
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp]:
```suggestion
  if(nullptr == fromGeometry)
```

File: Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+  {
+    QString ss = QObject::tr("Selected input DataContainer must contain either an Image Geometry or RectLinearGrid Geometry. The DataContainer did not contain a Geometry object.");
+    setErrorCondition(-2008, ss);
+    return;
+  }
+  IGeometry::Type geomType = fromGeometry->getGeometryType();
+  if(IGeometry::Type::Image == geomType || IGeometry::Type::RectGrid == geomType)
+  {
+    std::vector<size_t> cDims = {3};
+    m_CoordinatesPtr = getDataContainerArray()->createNonPrereqArrayFromPath<FloatArrayType>(this, getCoordinateArrayPath(), 0, cDims, "", DataArrayID31);
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp]:
```suggestion
    m_CoordinatesPtr = getDataContainerArray()->createNonPrereqArrayFromPath<FloatArrayType>(this, getCoordinateArrayPath(), 0, {3}, "", DataArrayID31);
```

File: Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+  {
+    return;
+  }
+
+  IGeometryGrid::Pointer sourceGeometry = getDataContainerArray()->getDataContainer(getSelectedDataContainerName())->getGeometryAs<IGeometryGrid>();
+
+  float coords[3] = {0.0f, 0.0f, 0.0f};
+
+  SizeVec3Type dims = sourceGeometry->getDimensions();
+  size_t cellCount = std::accumulate(dims.begin(), dims.end(), static_cast<size_t>(1), std::multiplies<size_t>());
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp:154]:
```suggestion
  size_t cellCount = std::accumulate(dims.cbegin(), dims.cend(), static_cast<size_t>(1), std::multiplies<size_t>());
```

File: Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+  initialize();
+  dataCheck();
+  if(getErrorCode() < 0)
+  {
+    return;
+  }
+
+  IGeometryGrid::Pointer sourceGeometry = getDataContainerArray()->getDataContainer(getSelectedDataContainerName())->getGeometryAs<IGeometryGrid>();
+
+  float coords[3] = {0.0f, 0.0f, 0.0f};
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp]:
```suggestion
  std::array<float, 3> coords = {0.0f, 0.0f, 0.0f};
```

File: Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+ *
+ * The code contained herein was partially funded by the followig contracts:
+ *    United States Air Force Prime Contract FA8650-07-D-5800
+ *    United States Air Force Prime Contract FA8650-10-D-5210
+ *    United States Prime Contract Navy N00173-07-C-2068
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+#include "GenerateVertexCoordinates.h"
+
+#include <numeric>
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp:38]:
```suggestion
#include <array>
#include <numeric>
```

File: Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+  SizeVec3Type dims = sourceGeometry->getDimensions();
+  size_t cellCount = std::accumulate(dims.begin(), dims.end(), static_cast<size_t>(1), std::multiplies<size_t>());
+
+  // Use the APIs from the IGeometryGrid to get the XYZ coord for the center of each cell and then set that into the
+  // the new VertexGeometry
+  for(size_t idx = 0; idx < cellCount; idx++)
+  {
+    sourceGeometry->getCoords(idx, coords);
+    m_CoordinatesPtr->setTuple(idx, coords);
```
Comment (JDuffeyBQ) [inline Source/SIMPLib/CoreFilters/GenerateVertexCoordinates.cpp]:
```suggestion
    sourceGeometry->getCoords(idx, coords.data());
    m_CoordinatesPtr->setTuple(idx, coords.data());
```

**Review by JDuffeyBQ — CHANGES_REQUESTED** [review]

**Comment by imikejackson** [comment]
https://github.com/BlueQuartzSoftware/DREAM3DCI/pull/63

**Review by JDuffeyBQ — APPROVED** [review]

**Pull request merged**